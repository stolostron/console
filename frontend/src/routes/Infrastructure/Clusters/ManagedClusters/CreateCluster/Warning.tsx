/* Copyright Contributors to the Open Cluster Management project */

import { Alert, AlertVariant, ButtonVariant } from '@patternfly/react-core'
import { AcmButton } from '../../../../../ui-components'
import { Link } from 'react-router-dom-v5-compat'
import { createContext, useContext } from 'react'

export type WarningContextType =
  | { title: string; text: string; linkTo?: string; linkText?: string; isExternalLink?: boolean }
  | undefined

export const WarningContext = createContext<WarningContextType>(undefined)

export const Warning = () => {
  const warning = useContext(WarningContext)

  if (!warning) {
    return null
  }

  let actionLinks
  if (warning.linkTo) {
    warning.isExternalLink
      ? (actionLinks = (
          <a href={warning.linkTo} target="_blank" rel="noopener noreferrer" style={{ paddingLeft: 0 }}>
            {warning.linkText}
          </a>
        ))
      : (actionLinks = (
          <Link to={warning.linkTo}>
            <AcmButton variant={ButtonVariant.link} style={{ paddingLeft: 0 }}>
              {warning.linkText}
            </AcmButton>
          </Link>
        ))
  }

  return (
    <>
      <Alert title={warning?.title} variant={AlertVariant.warning} actionLinks={actionLinks} isInline>
        {warning.text}
      </Alert>
      <br />
    </>
  )
}
