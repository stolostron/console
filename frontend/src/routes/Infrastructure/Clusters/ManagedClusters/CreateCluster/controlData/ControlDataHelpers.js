/* Copyright Contributors to the Open Cluster Management project */
import { VALIDATE_CIDR, VALIDATE_NUMERIC, VALIDATE_BASE_DNS_NAME_REQUIRED, VALID_DNS_LABEL } from 'temptifly'
import { listClusterImageSets } from '../../../../../../resources/cluster-image-set'
import { unpackProviderConnection } from '../../../../../../resources/provider-connection'
import { NavigationPath } from '../../../../../../NavigationPath'
import _ from 'lodash'

const OpenNewTab = () => (
    <svg width="24px" height="24px" x="0px" y="0px" viewBox="0 0 1024 1024" xmlSpace="preserve" role="presentation">
        <g stroke="none" strokeWidth="1" fillRule="evenodd">
            <path d="M576,320 L896,320 L896,192 L576,192 L576,320 Z M128,320 L448,320 L448,192 L128,192 L128,320 Z M930,64 L896,64 L128,64 L94,64 C42.085,64 0,106.085 0,158 L0,192 L0,832 L0,866 C0,917.915 42.085,960 94,960 L128,960 L488,960 C501.255,960 512,949.255 512,936 L512,856 C512,842.745 501.255,832 488,832 L140,832 C133.373,832 128,826.627 128,820 L128,448 L896,448 L896,552 C896,565.255 906.745,576 920,576 L1000,576 C1013.255,576 1024,565.255 1024,552 L1024,158 C1024,106.085 981.915,64 930,64 L930,64 Z" />
            <path d="M968,784 L848,784 L848,664 C848,650.7 837.3,640 824,640 L776,640 C762.7,640 752,650.7 752,664 L752,784 L632,784 C618.7,784 608,794.7 608,808 L608,856 C608,869.3 618.7,880 632,880 L752,880 L752,1000 C752,1013.3 762.7,1024 776,1024 L824,1024 C837.3,1024 848,1013.3 848,1000 L848,880 L968,880 C981.3,880 992,869.3 992,856 L992,808 C992,794.7 981.3,784 968,784" />
        </g>
    </svg>
)

export const CREATE_CLOUD_CONNECTION = {
    prompt: 'creation.ocp.cloud.add.connection',
    type: 'link',
    url: NavigationPath.addCredentials,
    positionBottomRight: true,
    id: 'add-provider-connection',
    icon: <OpenNewTab />,
}

export const CREATE_AUTOMATION_TEMPLATE = {
    prompt: 'creation.ocp.cloud.add.template',
    type: 'link',
    url: NavigationPath.addAnsibleAutomation,
    positionBottomRight: true,
    id: 'add-automation-template',
    icon: <OpenNewTab />,
}

export const LOAD_OCP_IMAGES = (provider) => {
    return {
        query: () => {
            return listClusterImageSets().promise
        },
        emptyDesc: 'creation.ocp.cloud.no.ocp.images',
        loadingDesc: 'creation.ocp.cloud.loading.ocp.images',
        setAvailable: setAvailableOCPImages.bind(null, provider),
        setAvailableMap: setAvailableOCPMap.bind(null),
    }
}

const getImageName = (image) => {
    return `img${image
        .split(':')
        .pop()
        .replace(/[^\w.]|_/g, '-')}`
}

const getImageVersion = (image) => {
    const match = /(\d+.\d+.\d+)-/gm.exec(image)
    return _.get(match, '1', '')
}

export const getSimplifiedImageName = (image) => {
    const match = /.+:(.*)-/gm.exec(image)
    if (match && match[1]) {
        return `OpenShift ${match[1]}`
    }
}

export const getWorkerName = (control) => {
    const { grpNum } = control
    return `Worker pool ${grpNum + 1}`
}

export const setAvailableOCPImages = (provider, control, result) => {
    const { loading } = result
    const { data } = result
    const imageSets = data
    control.isLoading = false
    const error = imageSets ? null : result.error
    if (!control.available) {
        control.available = []
        control.availableMap = {}
    }
    if (control.available.length === 0 && (error || imageSets)) {
        if (error) {
            control.hasReplacements = true
            control.noHandlebarReplacements = true
            control.isFailed = true
        } else if (imageSets) {
            control.isLoaded = true
            control.hasReplacements = true
            control.noHandlebarReplacements = true
            imageSets.forEach((item) => {
                const { metadata, spec } = item
                const name = metadata?.name
                const visible = metadata?.labels?.visible
                const releaseImage = spec?.releaseImage
                // We only hide when visible is false. We consider visible the default
                if (visible !== 'false') {
                    switch (provider) {
                        case 'bmc':
                            if (name.includes('img4.3')) {
                                return
                            }
                            break
                        case 'vmw':
                            if (name.includes('img4.3') || name.includes('img4.4')) {
                                return
                            }
                            break
                        default:
                            break
                    }
                    control.available.push(releaseImage)
                    control.availableMap[releaseImage] = {
                        replacements: {
                            releaseImageReference: name,
                            clusterImageSetComment: releaseImage,
                            releaseImageVersion: getImageVersion(name),
                        },
                    }
                }
            })
            control.available.reverse()
        }
    } else {
        control.isLoading = loading
    }
}

export const setAvailableOCPMap = (control) => {
    const { active, availableMap, isFailed } = control
    if (active && !availableMap[active]) {
        availableMap[active] = !isFailed
            ? {
                  replacements: {
                      clusterReleaseImage: active,
                      clusterImageSetName: getImageName(active),
                      releaseImageVersion: getImageVersion(active),
                  },
              }
            : {
                  replacements: {
                      releaseImage: active,
                  },
              }
    }
}

export const setAvailableConnections = (control, secrets) => {
    const connections = secrets.filter(
        (secret) => secret.metadata.labels?.['cluster.open-cluster-management.io/type'] === control.providerId
    )
    control.availableMap = {}
    connections?.forEach?.((c) => {
        const unpackedSecret = unpackProviderConnection(c)
        const replacements = { ...(unpackedSecret.stringData ?? {}) }
        if (replacements['sshKnownHosts']) {
            replacements['sshKnownHosts'] = replacements['sshKnownHosts'].split('\n')
        }
        // handlebars don't like periods in ids, so use a substitute tag - OpenStack
        if (replacements['clouds.yaml']) {
            replacements['cloudsYaml'] = replacements['clouds.yaml']
        }
        // handlebars don't like periods in ids, so use a substitute tag - Azure
        if (replacements['osServicePrincipal.json']) {
            replacements['osServicePrincipalJson'] = replacements['osServicePrincipal.json']
        }
        // handlebars don't like periods in ids, so use a substitute tag - GCP
        if (replacements['osServiceAccount.json']) {
            replacements['osServiceAccountJson'] = replacements['osServiceAccount.json']
        }
        control.availableMap[c.metadata.name] = { replacements }
        control.hasReplacements = true
        control.noHandlebarReplacements = true
        control.isLoaded = true
    })
    control.available = connections.map((secret) => secret.metadata.name)
}

export const setAvailableTemplates = (control, templates) => {
    control.available = templates.map((template) => template.metadata.name)
}

export const clusterDetailsControlData = [
    {
        id: 'detailStep',
        type: 'step',
        title: 'Cluster details',
    },
    {
        name: 'creation.ocp.name',
        tooltip: 'tooltip.creation.ocp.name',
        id: 'name',
        type: 'text',
        validation: {
            constraint: VALID_DNS_LABEL,
            notification: 'import.form.invalid.dns.label',
            required: true,
        },
        reverse: 'ClusterDeployment[0].metadata.name',
    },
    {
        name: 'creation.ocp.clusterSet',
        tooltip: 'tooltip.creation.ocp.clusterSet',
        id: 'clusterSet',
        type: 'singleselect',
        placeholder: 'placeholder.creation.ocp.clusterSet',
        validation: {
            required: false,
        },
        available: [],
    },
    {
        name: 'creation.ocp.baseDomain',
        tooltip: 'tooltip.creation.ocp.baseDomain',
        id: 'baseDomain',
        type: 'text',
        validation: VALIDATE_BASE_DNS_NAME_REQUIRED,
    },
]

export const networkingControlData = [
    ///////////////////////  networking  /////////////////////////////////////
    {
        id: 'networkType',
        name: 'creation.ocp.cluster.network.type',
        tooltip: 'tooltip.creation.ocp.cluster.network.type',
        type: 'singleselect',
        active: 'OpenShiftSDN',
        available: ['OpenShiftSDN', 'OVNKubernetes'],
        validation: {
            notification: 'creation.ocp.cluster.valid.key',
            required: true,
        },
    },
    {
        id: 'networks',
        type: 'group',
        prompts: {
            addPrompt: 'creation.ocp.cluster.add.network',
            deletePrompt: 'creation.ocp.cluster.delete.node.pool',
        },
        controlData: [
            {
                id: 'networkGroup',
                type: 'section',
                collapsable: true,
                subtitle: 'creation.ocp.node.network.title',
                info: 'creation.ocp.node.network.info',
            },
            {
                id: 'clusterNetwork',
                type: 'text',
                name: 'creation.ocp.cluster.network',
                tooltip: 'tooltip.creation.ocp.cluster.network',
                active: '10.128.0.0/14',
                validation: VALIDATE_CIDR,
            },
            {
                id: 'hostPrefix',
                type: 'text',
                name: 'creation.ocp.cluster.network.host.prefix',
                tooltip: 'tooltip.creation.ocp.cluster.network.host.prefix',
                active: '23',
                validation: VALIDATE_NUMERIC,
            },
            {
                id: 'serviceNetwork',
                type: 'text',
                name: 'creation.ocp.service.network',
                tooltip: 'tooltip.creation.ocp.service.network',
                active: '172.30.0.0/16',
                validation: VALIDATE_CIDR,
            },
            {
                id: 'machineCIDR',
                type: 'text',
                name: 'creation.ocp.machine.cidr',
                tooltip: 'tooltip.creation.ocp.machine.cidr',
                active: '10.0.0.0/16',
                validation: VALIDATE_CIDR,
            },
        ],
    },
]

export const automationControlData = [
    ///////////////////////  automation  /////////////////////////////////////
    {
        id: 'automationStep',
        type: 'step',
        title: 'template.clusterCreate.title',
    },
    {
        id: 'chooseTemplate',
        type: 'title',
        info: 'template.clusterCreate.info',
    },
    {
        name: 'template.clusterCreate.name',
        id: 'templateName',
        type: 'singleselect',
        placeholder: 'template.clusterCreate.select.placeholder',
        available: [],
        validation: {
            required: false,
        },
        prompts: CREATE_AUTOMATION_TEMPLATE,
    },
]

export const isHidden_lt_OCP48 = (control, controlData) => {
    const singleNodeFeatureFlag = controlData.find(({ id }) => id === 'singleNodeFeatureFlag')
    const imageSet = controlData.find(({ id }) => id === 'imageSet')
    //NOTE: We will need to adjust this in the future for new OCP versions!
    if (
        singleNodeFeatureFlag &&
        singleNodeFeatureFlag.active &&
        imageSet &&
        imageSet.active &&
        (imageSet.active.includes('release:4.8') ||
            imageSet.active.includes('release:4.9') ||
            imageSet.active.includes('release:4.10'))
    ) {
        return false
    }
    return true
}

export const isHidden_SNO = (control, controlData) => {
    const singleNode = controlData.find(({ id }) => id === 'singleNode')
    return singleNode && singleNode.active && !isHidden_lt_OCP48(control, controlData)
}

export const onChangeSNO = (control, controlData) => {
    var groupDataArray = controlData.find(({ id }) => id === 'workerPools').active
    groupDataArray.forEach((group) => {
        var computeNodeCount = group.find(({ id }) => id === 'computeNodeCount')
        if (!control.active) {
            if (computeNodeCount) {
                computeNodeCount.active = '3'
            }
        }
    })
}
