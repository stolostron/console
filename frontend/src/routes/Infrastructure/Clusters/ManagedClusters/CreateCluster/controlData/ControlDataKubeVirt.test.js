// Copyright (c) 2022 Red Hat, Inc.
// Copyright Contributors to the Open Cluster Management project
'use strict'

import i18next from 'i18next'
import { Warning } from '../Warning'
import { setAvailableStorageClasses } from './ControlDataHelpers'
import { getControlDataKubeVirt } from './ControlDataKubeVirt'

const t = i18next.t.bind(i18next)
const handleModalToggle = jest.fn()

describe('Cluster creation control data for KubeVirt', () => {
  const localCluster = {
    name: 'local-cluster',
    displayName: 'local-cluster',
    namespace: 'local-cluster',
    status: 'ready',
    provider: 'aws',
    labels: {
      cloud: 'Amazon',
      'cluster.open-cluster-management.io/clusterset': 'default',
      clusterID: '68438ab6-ceec-4f2f-9d9e-675b518ab0ef',
      'local-cluster': 'true',
      name: 'local-cluster',
      openshiftVersion: '4.13.23',
      'openshiftVersion-major': '4',
      'openshiftVersion-major-minor': '4.13',
      'velero.io/exclude-from-backup': 'true',
      vendor: 'OpenShift',
    },
    consoleURL: 'https://console-openshift-console.apps.cs-aws-413-j9kqv.dev02.red-chesterfield.com',
    isHive: false,
    isHypershift: false,
    isManaged: true,
    isCurator: false,
    hasAutomationTemplate: false,
    isHostedCluster: false,
    isSNOCluster: false,
    isRegionalHubCluster: false,
    hive: {
      isHibernatable: false,
      secrets: {},
    },
    clusterSet: 'default',
    owner: {},
    creationTimestamp: '2024-01-29T04:31:03Z',
  }

  it('generates correctly', () => {
    expect(getControlDataKubeVirt(t, handleModalToggle, <Warning />, true, {})).toMatchSnapshot()
  })
  it('generates correctly for MCE', () => {
    expect(getControlDataKubeVirt(t, handleModalToggle, <Warning />, false, localCluster)).toMatchSnapshot()
  })
  it('Correctly sets available storage classes', () => {
    const control = {
      controlId: 'storageClassName',
      isLoading: false,
      isLoaded: true,
      hasReplacements: true,
      availableMap: {
        'gp3-csi': {
          replacements: {
            storageClassName: 'gp3-csi',
          },
        },
      },
      available: [],
    }
    const result = {
      loading: false,
      data: [
        {
          metadata: {
            name: 'storageclass1',
          },
          provisioner: 'test.com',
          parameters: {
            encrypted: 'true',
            type: 'sc1',
          },
          reclaimPolicy: 'Delete',
          allowVolumeExpansion: true,
          volumeBindingMode: 'WaitForFirstConsumer',
          apiVersion: 'storage.k8s.io/v1',
          kind: 'StorageClass',
        },
        {
          metadata: {
            name: 'storageclass2',
            annotations: {
              'storageclass.kubernetes.io/is-default-class': 'true',
            },
          },
          provisioner: 'test.com',
          parameters: {
            encrypted: 'true',
            type: 'sc2',
          },
          reclaimPolicy: 'Delete',
          allowVolumeExpansion: true,
          volumeBindingMode: 'WaitForFirstConsumer',
          apiVersion: 'storage.k8s.io/v1',
          kind: 'StorageClass',
        },
      ],
    }
    setAvailableStorageClasses(control, result)
    expect(control).toMatchSnapshot()
  })
})
