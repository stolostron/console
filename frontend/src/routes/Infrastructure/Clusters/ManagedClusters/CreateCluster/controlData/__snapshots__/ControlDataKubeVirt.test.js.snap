// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cluster creation control data for KubeVirt Correctly sets available storage classes 1`] = `
Object {
  "available": Array [
    "storageclass2",
  ],
  "availableMap": Object {
    "gp3-csi": Object {
      "replacements": Object {
        "storageClassName": "gp3-csi",
      },
    },
    "storageclass2": Object {
      "replacements": Object {
        "storageClassName": "storageclass2",
      },
    },
  },
  "controlId": "storageClassName",
  "hasReplacements": true,
  "isLoaded": true,
  "isLoading": false,
  "noHandlebarReplacements": true,
}
`;

exports[`Cluster creation control data for KubeVirt generates correctly 1`] = `
Array [
  Object {
    "id": "kubevirtDetailStep",
    "title": "Cluster details",
    "type": "step",
  },
  Object {
    "active": "OpenShift Virtualization",
    "id": "infrastructure",
    "name": "Infrastructure",
    "type": "reviewinfo",
  },
  Object {
    "active": "Hosted",
    "id": "controlplane",
    "name": "Control plane type",
    "type": "reviewinfo",
  },
  Object {
    "component": <Alert
      isInline={true}
      style={
        Object {
          "marginBottom": "1rem",
        }
      }
      title="Operator required"
      variant="danger"
    >
      <div
        style={
          Object {
            "display": "flex",
            "flexDirection": "column",
          }
        }
      >
        <span
          style={
            Object {
              "margin": "0.5rem 0",
            }
          }
        >
          OpenShift Virtualization operator is required to create a cluster.
        </span>
        <AcmButton
          component="a"
          href=""
          icon={
            <ExternalLinkAltIcon
              color="currentColor"
              noVerticalAlign={false}
              size="sm"
            />
          }
          iconPosition="right"
          isInline={true}
          target="_blank"
          variant="link"
        >
          Install operator
        </AcmButton>
      </div>
    </Alert>,
    "hidden": false,
    "id": "kubevirt-operator-alert",
    "type": "custom",
  },
  Object {
    "available": Array [],
    "footer": <CreateCredentialModal
      handleModalToggle={[MockFunction]}
    />,
    "hasReplacements": true,
    "id": "connection",
    "name": "Infrastructure provider credential",
    "onSelect": [Function],
    "placeholder": "Select a credential",
    "providerId": "kubevirt",
    "tooltip": "The settings that are required for the selected provider. You can select an existing credential, or add new credentials. These credentials contain values such as pull secret or cloud credentials. The selected credential cannot be changed after creation.",
    "type": "singleselect",
    "validation": Object {
      "notification": "Select a connection",
      "required": true,
    },
  },
  Object {
    "active": true,
    "id": "showSecrets",
    "type": "hidden",
  },
  Object {
    "id": "clusterName",
    "name": "Cluster name",
    "placeholder": "Enter cluster name",
    "tooltip": "The unique name of your cluster. The value must be a string that contains lowercase alphanumeric values, such as dev. Cannot be changed after creation.",
    "type": "text",
    "validation": Object {
      "constraint": "^[a-z0-9]([a-z0-9-]{0,61}[a-z0-9])?$",
      "notification": "The value must be a valid DNS label that consists of up to 63 lowercase alphanumeric characters. The character '-' is also permitted, as long as it does not appear in the first or last position.",
      "required": true,
    },
  },
  Object {
    "available": Array [],
    "id": "clusterSet",
    "name": "Cluster set",
    "placeholder": "Select a cluster set",
    "tooltip": "A ManagedClusterSet is a group of managed clusters. With a ManagedClusterSet, you can manage access to all of the managed clusters in the group together. Clusters from the pool will be added to this set. After the cluster pool is created, the cluster set cannot be changed.",
    "type": "singleselect",
    "validation": Object {
      "required": false,
    },
  },
  Object {
    "fetchAvailable": Object {
      "emptyDesc": "No default images. Enter a release image path.",
      "loadingDesc": "Loading release image sets...",
      "query": [Function],
      "setAvailable": [Function],
      "setAvailableMap": [Function],
    },
    "id": "releaseImage",
    "name": "Release image",
    "onSelect": [Function],
    "placeholder": "Select or enter a release image",
    "reverse": [Function],
    "simplified": [Function],
    "tooltip": "URL to the OpenShift install image set to use. Available images are listed, or you can enter your own path to add an image to the list.",
    "type": "combobox",
    "validation": Object {
      "notification": "Select a release image",
      "required": true,
    },
  },
  Object {
    "fetchAvailable": Object {
      "emptyDesc": "No etcd storage classes. Enter an etcd storage classes name.",
      "loadingDesc": "Loading etcd storage classes...",
      "query": [Function],
      "setAvailable": [Function],
    },
    "id": "storageClassName",
    "name": "Etcd storage class",
    "placeholder": "Select etcd storage class",
    "reverse": [Function],
    "tooltip": "Persistent volume storage class for etcd data volumes",
    "type": "combobox",
    "validation": Object {
      "notification": "Select a storage class",
      "required": false,
    },
  },
  Object {
    "active": Object {
      "controller": "HighlyAvailable",
      "infra": "HighlyAvailable",
    },
    "component": <AvailabilityOptionsForm />,
    "id": "availabilityOptions",
    "summary": [Function],
    "type": "custom",
  },
  Object {
    "active": Array [],
    "id": "additionalLabels",
    "name": "Additional labels",
    "tooltip": "Use labels to organize and place application subscriptions and policies on this cluster. The placement of resources are controlled by label selectors. If your cluster has the labels that match the resource placement’s label selector, the resource will be installed on your cluster after creation.",
    "type": "labels",
  },
  Object {
    "id": "nodepoolsStep",
    "title": "Node pools",
    "type": "step",
  },
  Object {
    "controlData": Array [
      Object {
        "collapsable": true,
        "id": "nodepoolsGroup",
        "subtitle": [Function],
        "type": "section",
      },
      Object {
        "id": "nodePoolName",
        "name": "Node pool name",
        "placeholder": "Enter node pool name",
        "tooltip": "The unique name of your node pool. The value must be a string that contains lowercase alphanumeric values, such as dev. Cannot be changed after creation.",
        "type": "text",
        "validation": Object {
          "constraint": "^[a-z0-9]([a-z0-9-]{0,61}[a-z0-9])?$",
          "notification": "The value must be a valid DNS label that consists of up to 63 lowercase alphanumeric characters. The character '-' is also permitted, as long as it does not appear in the first or last position.",
          "required": true,
        },
      },
      Object {
        "id": "nodePoolReplica",
        "initial": "2",
        "name": "Node pool replica",
        "tooltip": "The number of Worker Nodes in the cluster",
        "type": "number",
        "validation": Object {
          "notification": "Value must be integer.",
          "required": true,
          "tester": /\\^\\\\d\\+\\$/,
        },
      },
      Object {
        "id": "nodePoolCoreCount",
        "initial": "2",
        "min": 1,
        "name": "Core",
        "tooltip": "Number of cores assigned to each worker node VM",
        "type": "number",
      },
      Object {
        "id": "nodePoolMemory",
        "initial": "8",
        "min": 1,
        "name": "Memory (GiB)",
        "tooltip": "Memory assigned to each worker node VM",
        "type": "number",
      },
      Object {
        "id": "nodePoolAutoRepair",
        "isTrue": false,
        "name": "Auto repair",
        "tooltip": "Enables machine auto-repair with machine health checks",
        "type": "boolean",
      },
      Object {
        "collapsable": true,
        "collapsed": true,
        "id": "rootVolumeSection",
        "title": "Root volume option",
        "type": "section",
      },
      Object {
        "id": "rootVolumeSize",
        "initial": "32",
        "name": "Size (GiB)",
        "tooltip": "The size of the underlying root volume used for the worker node VM",
        "type": "number",
      },
      Object {
        "id": "rootVolumeStorageClass",
        "name": "Root Volume Storage Class",
        "placeholder": "Enter Storage Class",
        "tooltip": "Storage class used for the worker node VM’s root volume",
        "type": "text",
      },
      Object {
        "available": Array [
          "ReadWriteOnce",
          "ReadWriteMany",
          "ReadOnly",
          "ReadWriteOncePod",
        ],
        "id": "rootVolumeAccessMode",
        "name": "Access mode",
        "placeholder": "Select an access mode",
        "tooltip": "Access mode used for the worker node VM’s root volume",
        "type": "combobox",
      },
      Object {
        "available": Array [
          "Block",
          "Filesystem",
        ],
        "id": "rootVolumeVolMode",
        "name": "Volume mode",
        "placeholder": "Select a volume mode",
        "tooltip": "Volume Mode used for the worker node VM’s root volume",
        "type": "combobox",
      },
    ],
    "id": "nodepools",
    "prompts": Object {
      "addPrompt": "Add node pool",
      "deletePrompt": "Delete node pool",
    },
    "type": "group",
  },
  Object {
    "active": true,
    "id": "includeKlusterletAddonConfig",
    "type": "hidden",
  },
  Object {
    "component": <Warning />,
    "id": "warning",
    "type": "custom",
  },
]
`;

exports[`Cluster creation control data for KubeVirt generates correctly for MCE 1`] = `
Array [
  Object {
    "id": "kubevirtDetailStep",
    "title": "Cluster details",
    "type": "step",
  },
  Object {
    "active": "OpenShift Virtualization",
    "id": "infrastructure",
    "name": "Infrastructure",
    "type": "reviewinfo",
  },
  Object {
    "active": "Hosted",
    "id": "controlplane",
    "name": "Control plane type",
    "type": "reviewinfo",
  },
  Object {
    "component": <Alert
      isInline={true}
      style={
        Object {
          "marginBottom": "1rem",
        }
      }
      title="Operator required"
      variant="danger"
    >
      <div
        style={
          Object {
            "display": "flex",
            "flexDirection": "column",
          }
        }
      >
        <span
          style={
            Object {
              "margin": "0.5rem 0",
            }
          }
        >
          OpenShift Virtualization operator is required to create a cluster.
        </span>
        <AcmButton
          component="a"
          href="https://console-openshift-console.apps.cs-aws-413-j9kqv.dev02.red-chesterfield.com/operatorhub/all-namespaces?keyword=Openshift+Virtualization"
          icon={
            <ExternalLinkAltIcon
              color="currentColor"
              noVerticalAlign={false}
              size="sm"
            />
          }
          iconPosition="right"
          isInline={true}
          target="_blank"
          variant="link"
        >
          Install operator
        </AcmButton>
      </div>
    </Alert>,
    "hidden": false,
    "id": "kubevirt-operator-alert",
    "type": "custom",
  },
  Object {
    "available": Array [],
    "footer": <CreateCredentialModal
      handleModalToggle={[MockFunction]}
    />,
    "hasReplacements": true,
    "id": "connection",
    "name": "Infrastructure provider credential",
    "onSelect": [Function],
    "placeholder": "Select a credential",
    "providerId": "kubevirt",
    "tooltip": "The settings that are required for the selected provider. You can select an existing credential, or add new credentials. These credentials contain values such as pull secret or cloud credentials. The selected credential cannot be changed after creation.",
    "type": "singleselect",
    "validation": Object {
      "notification": "Select a connection",
      "required": true,
    },
  },
  Object {
    "active": true,
    "id": "showSecrets",
    "type": "hidden",
  },
  Object {
    "id": "clusterName",
    "name": "Cluster name",
    "placeholder": "Enter cluster name",
    "tooltip": "The unique name of your cluster. The value must be a string that contains lowercase alphanumeric values, such as dev. Cannot be changed after creation.",
    "type": "text",
    "validation": Object {
      "constraint": "^[a-z0-9]([a-z0-9-]{0,61}[a-z0-9])?$",
      "notification": "The value must be a valid DNS label that consists of up to 63 lowercase alphanumeric characters. The character '-' is also permitted, as long as it does not appear in the first or last position.",
      "required": true,
    },
  },
  Object {
    "available": Array [],
    "id": "clusterSet",
    "name": "Cluster set",
    "placeholder": "Select a cluster set",
    "tooltip": "A ManagedClusterSet is a group of managed clusters. With a ManagedClusterSet, you can manage access to all of the managed clusters in the group together. Clusters from the pool will be added to this set. After the cluster pool is created, the cluster set cannot be changed.",
    "type": "singleselect",
    "validation": Object {
      "required": false,
    },
  },
  Object {
    "fetchAvailable": Object {
      "emptyDesc": "No default images. Enter a release image path.",
      "loadingDesc": "Loading release image sets...",
      "query": [Function],
      "setAvailable": [Function],
      "setAvailableMap": [Function],
    },
    "id": "releaseImage",
    "name": "Release image",
    "onSelect": [Function],
    "placeholder": "Select or enter a release image",
    "reverse": [Function],
    "simplified": [Function],
    "tooltip": "URL to the OpenShift install image set to use. Available images are listed, or you can enter your own path to add an image to the list.",
    "type": "combobox",
    "validation": Object {
      "notification": "Select a release image",
      "required": true,
    },
  },
  Object {
    "fetchAvailable": Object {
      "emptyDesc": "No etcd storage classes. Enter an etcd storage classes name.",
      "loadingDesc": "Loading etcd storage classes...",
      "query": [Function],
      "setAvailable": [Function],
    },
    "id": "storageClassName",
    "name": "Etcd storage class",
    "placeholder": "Select etcd storage class",
    "reverse": [Function],
    "tooltip": "Persistent volume storage class for etcd data volumes",
    "type": "combobox",
    "validation": Object {
      "notification": "Select a storage class",
      "required": false,
    },
  },
  Object {
    "active": Object {
      "controller": "HighlyAvailable",
      "infra": "HighlyAvailable",
    },
    "component": <AvailabilityOptionsForm />,
    "id": "availabilityOptions",
    "summary": [Function],
    "type": "custom",
  },
  Object {
    "active": Array [],
    "id": "additionalLabels",
    "name": "Additional labels",
    "tooltip": "Use labels to organize and place application subscriptions and policies on this cluster. The placement of resources are controlled by label selectors. If your cluster has the labels that match the resource placement’s label selector, the resource will be installed on your cluster after creation.",
    "type": "labels",
  },
  Object {
    "id": "nodepoolsStep",
    "title": "Node pools",
    "type": "step",
  },
  Object {
    "controlData": Array [
      Object {
        "collapsable": true,
        "id": "nodepoolsGroup",
        "subtitle": [Function],
        "type": "section",
      },
      Object {
        "id": "nodePoolName",
        "name": "Node pool name",
        "placeholder": "Enter node pool name",
        "tooltip": "The unique name of your node pool. The value must be a string that contains lowercase alphanumeric values, such as dev. Cannot be changed after creation.",
        "type": "text",
        "validation": Object {
          "constraint": "^[a-z0-9]([a-z0-9-]{0,61}[a-z0-9])?$",
          "notification": "The value must be a valid DNS label that consists of up to 63 lowercase alphanumeric characters. The character '-' is also permitted, as long as it does not appear in the first or last position.",
          "required": true,
        },
      },
      Object {
        "id": "nodePoolReplica",
        "initial": "2",
        "name": "Node pool replica",
        "tooltip": "The number of Worker Nodes in the cluster",
        "type": "number",
        "validation": Object {
          "notification": "Value must be integer.",
          "required": true,
          "tester": /\\^\\\\d\\+\\$/,
        },
      },
      Object {
        "id": "nodePoolCoreCount",
        "initial": "2",
        "min": 1,
        "name": "Core",
        "tooltip": "Number of cores assigned to each worker node VM",
        "type": "number",
      },
      Object {
        "id": "nodePoolMemory",
        "initial": "8",
        "min": 1,
        "name": "Memory (GiB)",
        "tooltip": "Memory assigned to each worker node VM",
        "type": "number",
      },
      Object {
        "id": "nodePoolAutoRepair",
        "isTrue": false,
        "name": "Auto repair",
        "tooltip": "Enables machine auto-repair with machine health checks",
        "type": "boolean",
      },
      Object {
        "collapsable": true,
        "collapsed": true,
        "id": "rootVolumeSection",
        "title": "Root volume option",
        "type": "section",
      },
      Object {
        "id": "rootVolumeSize",
        "initial": "32",
        "name": "Size (GiB)",
        "tooltip": "The size of the underlying root volume used for the worker node VM",
        "type": "number",
      },
      Object {
        "id": "rootVolumeStorageClass",
        "name": "Root Volume Storage Class",
        "placeholder": "Enter Storage Class",
        "tooltip": "Storage class used for the worker node VM’s root volume",
        "type": "text",
      },
      Object {
        "available": Array [
          "ReadWriteOnce",
          "ReadWriteMany",
          "ReadOnly",
          "ReadWriteOncePod",
        ],
        "id": "rootVolumeAccessMode",
        "name": "Access mode",
        "placeholder": "Select an access mode",
        "tooltip": "Access mode used for the worker node VM’s root volume",
        "type": "combobox",
      },
      Object {
        "available": Array [
          "Block",
          "Filesystem",
        ],
        "id": "rootVolumeVolMode",
        "name": "Volume mode",
        "placeholder": "Select a volume mode",
        "tooltip": "Volume Mode used for the worker node VM’s root volume",
        "type": "combobox",
      },
    ],
    "id": "nodepools",
    "prompts": Object {
      "addPrompt": "Add node pool",
      "deletePrompt": "Delete node pool",
    },
    "type": "group",
  },
  Object {
    "active": false,
    "id": "includeKlusterletAddonConfig",
    "type": "hidden",
  },
  Object {
    "component": <Warning />,
    "id": "warning",
    "type": "custom",
  },
]
`;
